komodo/pypi_dependencies.py

```diff
@@ -0,0 +1,234 @@
+                    dist = pkginfo.Wheel(os.path.join(tmpdir, file))
+                else:
+                    dist = pkginfo.SDist(os.path.join(tmpdir, file))
+                self.requirements[canonical][package_version] = dist.requires_dist
+
+        return [Requirement(r) for r in self.requirements[canonical][package_version]]
+
+    def _make_install_name(self, name: str) -> str:
+        canonical = canonicalize_name(name)
+        return self._install_names.get(canonical, canonical)
+
+    def _version_satisfied(self, version: str, requirement: Requirement) -> bool:
+        if version in ["main", "master", LATEST_PACKAGE_ALIAS]:
+            return True
+        try:
+            specifier = requirement.specifier
+            specifier.prereleases = True
+            return version in specifier
+        except InvalidVersion:
+            return True
+
+    def _is_necessary(self, requirement, extras):
+        environment["extra"] = ",".join(extras)
+        return requirement.marker is None or requirement.marker.evaluate(environment)
+
+    def _satisfied(
+        self,
+        requirements: List[Requirement],
+    ) -> Tuple[bool, List[Tuple[List[Requirement], Set[str]]]]:
+        installed = self._to_install
+        also = []
+        for r in requirements:
+            name = canonicalize_name(r.name)
+            if name not in installed:
+                print(f"Not installed: {name}")
+                return False, []
+            self.used_packages.add(self._install_names[name])
+            installed_version = installed[name]
+
+            if not self._version_satisfied(installed_version, r):
+                return False, []
+
+            also.append((self.get_requirements(r.name, installed_version), r.extras))
+        return True, also
+
+    def satisfied(
+        self,
+        requirement: Requirement,
+        extra=None,
+    ) -> bool:
+        """
+        >>> from packaging.requirements import Requirement
+        >>> PypiDependencies(to_install={}, python_version="3.8").satisfied(
+        ... Requirement("PyJWT[crypto] (<3, >=1.0.0)")
+        ... )
+        Not installed: pyjwt
+        False
+        >>> PypiDependencies(to_install={"PyJWT": "2.3.0"}, python_version="3.8").satisfied(
+        ...     Requirement("PyJWT (<3, >=1.0.0)")
+        ... )
+        True
+        """
+        extra = extra or set()
+        if not self._is_necessary(requirement, extra):
+            return True
+        if requirement in self.failed_requirements:
+            return False
+        if requirement in self.satisfied_requirements:
+            return True
+
+        satisfied, also = self._satisfied([requirement])
+        if satisfied:
+            self.satisfied_requirements.add(requirement)
+        else:
+            self.failed_requirements.add(requirement)
+        return satisfied and all(
+            [self.satisfied(a, extra) for (rs, extra) in also for a in rs]
+        )
```

Comment:
Personally I'd go for something different than also and extra, but could you consider replacing `a`,  and `rs` with something more meaningful here?

Reply:
Tried to rename it to transient_requirements